addons/binding/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/channel/AbstractKNXChannelType.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.knx.internal.channel;
+
+import static java.util.stream.Collectors.*;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.types.Type;
+import org.openhab.binding.knx.KNXTypeMapper;
+import org.openhab.binding.knx.client.InboundSpec;
+import org.openhab.binding.knx.client.OutboundSpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tuwien.auto.calimero.GroupAddress;
+import tuwien.auto.calimero.KNXFormatException;
+
+/**
+ * Abstract base class for the meta-data abstraction for the KNX channel configurations.
+ *
+ * @author Karel Goderis - Initial contribution
+ */
+public abstract class AbstractKNXChannelType {
+
+    protected final Set<String> channelTypeIDs;
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractKNXChannelType.class);
+
+    protected AbstractKNXChannelType(String... channelTypeIDs) {
+        this.channelTypeIDs = new HashSet<>(Arrays.asList(channelTypeIDs));
+    }
+
+    @Nullable
+    protected abstract ChannelConfiguration parse(@Nullable String fancy);
+
+    public final Set<String> getChannelIDs() {
+        return channelTypeIDs;
+    }
+
+    protected final Set<String> asSet(String... values) {
+        return Collections.unmodifiableSet(new HashSet<>(Arrays.asList(values)));
+    }
+
+    protected abstract @NonNull Set<@NonNull String> getAllGAKeys();
```

Comment:
You should annotate the class with `@NonNullByDefault` and use `@Nullable` where needed. 

See also the [ESH documentation](https://www.eclipse.org/smarthome/documentation/development/conventions.html#null-annotations).